core/src/main/java/io/parsingdata/metal/expression/value/reference/CurrentIteration.java

```diff
@@ -73,12 +92,35 @@ public String toString() {
 
     @Override
     public boolean equals(final Object obj) {
-        return Util.notNullAndSameClass(this, obj);
+        return Util.notNullAndSameClass(this, obj)
+                && Objects.equals(level, ((CurrentIteration)obj).level);
     }
 
     @Override
     public int hashCode() {
-        return getClass().hashCode();
+        return Objects.hash(getClass(), level);
+    }
+
+    /**
+     * ParseGraphCandidates keeps track of the last n ParseGraphs.
+     */
+    private class ParseGraphCandidates {
+
+        private final Queue<ParseGraph> list;
+
+        ParseGraphCandidates(final int n) {
+            list = new ArrayDeque<>(nCopies(n, ParseGraph.EMPTY));
```

Comment:
Hmm, this is mutable state. Could you return a new instance of this class when adding an element? 

I do not know how others feel about using a stlib ArrayDeque in metal (@jvdb ?), an ImmutableList could also be used instead.